Implement JavaInstanceSpecLoader for exonum-launcher [ECR-3807] (#1226)
exonum/exonum-java-binding: exonum-java-binding/launcher-plugins/README.md

# Java Runtime Plugins for [Exonum Launcher](https://github.com/exonum/exonum-launcher)

- Java Runtime Plugin allows to deploy and start Java services.

> - Java Instance Plugin allows to use an arbitrary Protobuf message for service 

Reviewer: Local Configuration Plugin? Instance Configuration Plugin? (Instance) Initial Configuration Plugin?

It is not Java-specific, as it says below.

Author: Should I also rename the python package name?

Reviewer: If that makes sense — please do.

Author: Kinda hard to decide what name suits better. @popzxc @skletsun @dip56 what do you think?

Reviewer 2: I have no particular opinion on this topic.

  initial configuration parameters. The plugin can be used for both Java and Rust services.

## Installation

#### Requirements

- [Python3](https://www.python.org/downloads/)

```bash
# Generate sources
mvn generate-sources -pl core
# Install plugins
python3 -m pip install -e launcher-plugins
```

## Usage

TODO: Move the entire section to the EJB App tutorial.

To deploy and start a specific list of services, use the following command with a
prepared `config.yml` file:

```bash
python3 -m exonum_launcher -i config.yml
```

### Writing Configuration File

Start with specifying IP addresses of the blockchain nodes:

```yaml
networks:
  - host: "127.0.0.1"
    ssl: false
    public-api-port: 8080
    private-api-port: 8081
```

> You need to specify every node in the network.

Reviewer: Does it require the private ports of nodes to be accessible from a single location? I wonder if there are any other operation on the private port that do not authorize the user, making such setup impractical in distributed scenarious. Is that a known, deliberate restriction?

Author: ping @popzxc

Reviewer 2: `exonum-launcher` provides functionality to interact with nodes that you own. If you own all the nodes, you can send all the requests from one launch.
If network has several administrators, every administrator has to run `exonum-launcher` with the networks he owns.

Decentralized supervisor (as the oppose to the simple supervisor, which requires 1 request only, and which is now default) requires N requests and `exonum-launcher` is *trying* to make the life of the developer better, but this is not a generic solution for every situation.

Reviewer: Thanks for clarifications!

@vitvakatu , The documentation seems incomplete and has to be updated then, doesn't it?

Author: @dmitry-timofeev it seems, I will update.

Deadline height describes the maximum blockchain height for deployment process. Make sure to
specify the value larger than current blockchain height.

```yaml
deadline_height: 20000
```

Enable Java runtime by specifying its identifier (`1`). Rust runtime is enabled by default:

```yaml
runtimes:
  java: 1
```

Add artifacts to be deployed. For each artifact, you need to specify its name alias (as YAML key) 
and its runtime (using `runtime` field). Name aliases are used in other parts of configuration
for readability and easier refactoring. Java artifacts also need name of the `jar` file in 
`spec: artifact_filename` field. In our example we add Java `cryptocurrency-demo` service, and two
Rust services - `timestamping` and `time` oracle service.

```yaml
artifacts:
  cryptocurrency:
    runtime: java
    name: "com.exonum.examples:cryptocurrency:0.9.0-SNAPSHOT"
    spec:
      artifact_filename: "exonum-java-binding-cryptocurrency-demo-0.9.0-SNAPSHOT-artifact.jar"
  time:
    runtime: rust
    name: "exonum-time:0.12.0"
  timestamping:
    runtime: rust
    name: "exonum-timestamping:0.12.0"
```

> Add `plugins` section to enable both Java Runtime plugin and Java Instance plugin. Runtime plugin is

Reviewer 3: ```suggestion
Add a `plugins` section to enable both Java Runtime plugin and Java Instance plugin. Runtime plugin is
```

enabled for a specific runtime (`java` in our example), while Instance plugin is enabled for a
specific artifact name alias (`timestamping` in our example).

```yaml
plugins:
  runtime:
    java: "exonum_java_runtime_plugin.JavaDeploySpecLoader"
  artifact:
    timestamping: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
```

> In our example we will use Java Instance plugin to serialize initial configuration parameters of

Reviewer 3: ```suggestion
In our example we will use the Java Instance plugin to serialize initial configuration parameters of
```

> `timestamping` service in Protobuf. We need to take a `service.proto` file with message

Reviewer 3: ```suggestion
the `timestamping` service in Protobuf. We need to take a `service.proto` file with the message
```

> description from service sources and place it inside some known directory.

Reviewer 3: ```suggestion
description from the service sources and place it inside some known directory.
```

  ```proto
  syntax = "proto3";
  
  package exonum.examples.timestamping;
  
  message Config {
      string time_service_name = 1;
  }
  ```

> Finally, add `instances` section that describes the list of service instances to be started in the

Reviewer 3: ```suggestion
Finally, add an `instances` section that describes the list of service instances you want to start in the
```

> blockchain. For each instance you need to specify artifact name alias. Java Instance plugin also

Reviewer 3: ```suggestion
blockchain. For each instance you need to specify its artifact name alias. Java Instance plugin also
```

> requires a list of additional parameters, which we provide for `timestamping` instance:

Reviewer 3: ```suggestion
requires a list of additional parameters, which we provide for the `timestamping` instance:
```

> - `sources`. Points to a directory with Protobuf-sources of service configuration 

Reviewer 3: ```suggestion
- `sources`. Points to a directory with the Protobuf sources of the service configuration 
```

> message. We use `proto_sources` directory.

Reviewer 3: ```suggestion
message. We use the `proto_sources` directory.
```

> - `config_message_source`. A file name where `message_name` message 

Reviewer 3: ```suggestion
- `config_message_source`. A file name where the `message_name` message 
```

> is located. In our example we use `service.proto` file.

Reviewer 3: ```suggestion
is located. In our example we use the `service.proto` file.
```

> - `message_name`. A name of the Protobuf message used to represent service configuration.

Reviewer 3: ```suggestion
- `message_name`. A name of the Protobuf message used to represent the service configuration.
```

  Optional, defaults to `Config`.

> - `data`. Your actual configuration in the format corresponding to `message_name` message.

Reviewer 3: ```suggestion
- `data`. Your actual configuration in the format corresponding to the `message_name` message.
```

```yaml
instances:
  cryptocurrency:
    artifact: cryptocurrency
  time:
    artifact: time
  timestamping:
    artifact: timestamping
    config:
      sources: "proto_sources"
      config_message_source: "service.proto"
      message_name: "Config"
      data:
        time_service_name: "time"
```

> See [sample-config.yml](sample-config.yml) for the final state of configuration file.

Reviewer 3: ```suggestion
See [sample-config.yml](sample-config.yml) for the final state of the configuration file.
```

# Java Runtime Plugins for Exonum Launcher

## Installation

#### Requirements

- [Python3](https://www.python.org/downloads/)

```bash
# Generate sources
mvn generate-sources -pl core
# Install plugin
python3 -m pip install -e launcher-plugins
```

## Usage

TODO: Move the entire section to the EJB App tutorial.

### Java Runtime Plugin

Add `plugins` session to configuration file of the Exonum Launcher:

```yaml
runtimes:
  java: 1

plugins:
  runtime:
    java: "exonum_java_runtime_plugin.JavaDeploySpecLoader"
  artifact: {}
```

To load service artifact, provide a filename of the service artifact, for example:

```yaml
artifacts:
  cryptocurrency:
    runtime: java
    name: "com.exonum.examples:cryptocurrency:0.9.0-SNAPSHOT"
    spec:
      artifact_filename: "cryptocurrency-0.9.0-SNAPSHOT-artifact.jar"
```

### Java Instance Plugin

Add `plugins` session to configuration file of the Exonum Launcher:

```yaml
runtimes:
  java: 1

plugins:
  runtime: {}
  artifact: 

>     service_instance_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Reviewer: service_instance_name? What do service instances have to do with the plugin?

Author: My mistake, it must be `service_artifact_name`, to bind a specific plugin to a specific artifact.

```

> To instantiate a service with a custom configuration you need to create Protobuf

Reviewer: create — I think, _take_ a service configuration message protobuf definition (as it is defined by the service and must be used as is).

source of the configuration message and place it in specific directory. The name 

> of the message must be `Config`:

Reviewer: Is it a requirement? Or convention-over-configuration thing? The QA service currently uses the InitialConfiguration name.

We currently don't support re-configuration, but that interface is allowed to have a different message for configuration updates (though I don't think it is very likely). Will the need to support two different 'configuration' messages affect this restriction?

Author: It's possible to relax such restriction, thanks to Python

  ```proto
  syntax = "proto3";
  
  package exonum.examples.service_name;
  
  message Config {
      string time_service_name = 1;
  }
  ```

To instantiate a service, add the following fields to the instance `config`:

- `sources`. Points to a directory with Protobuf-sources of service configuration 
message. We use `proto_sources` directory.

> - `module_name`. A name (without extension) of the file where `Config` message 

Reviewer: Would a full filename as 'config_message_source' or 'config_filename' be easier to understand than a 'module_name'?

Author: Probably yes, naming suggestions are always welcomed

is located. In our example we use `service.proto` file.
- `data`. Your actual configuration in the format corresponding to `Config` message.

```yaml
instances:

>   service_instance_name:

Reviewer: Why does 'service_instance_name' have anything but a string service instance name?

Reviewer: Or is the key = instance name?

Author: Yep, the key is the instance name, and the value is additional parameters

# Java Runtime Plugins for Exonum Launcher

## Installation

#### Requirements

- [Python3](https://www.python.org/downloads/)

```bash
# Generate sources
mvn generate-sources -pl core
# Install plugin
python3 -m pip install -e launcher-plugins
```

## Usage

> TODO: Move the entire section to the EJB App tutorial.

Reviewer 4: Is this going to be implemented within this PR?

Author: Probably not, the entire readme isn't even a part of this task.

plugins:
  runtime: {}
  artifact: 

>     service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Author: The key is the artifact identifier from `artifacts` section of the config.
```yaml
artifacts:
  service_artifact_name:
    runtime: java
    name: "com.exonum.examples:service-name:0.9.0-SNAPSHOT"
    spec:
      artifact_filename: "service-artifact-filename.jar"
```

Added a sample config and will probably try to rephrase readme cause it's not obvious for reader.

>     service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Reviewer: Ok, so this format requires a second name for the artifact (in the example config, same as instance names). Is it worth it to add a second, shorter name instead of using the name an artifact already has?

>     service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Author: ping @popzxc 
The plugin does not change this part of the exonum-launcher. Though such implementation should be possible, as YAML supports strings as keys.

>     service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Reviewer 2: Plugins section has the following format:
```yaml
plugins:
  runtime:
     $RUNTIME_NAME: $RUNTIME_PLUGIN_PYTHON_CLASS
  artifact:
    $SPEC_ARTIFACT_ALIAS: $SPEC_PLUGIN_PYTHON_CLASS
```

So, for example let's take this config:
```yaml
artifacts:
  cryptocurrency:
    runtime: sample_runtime
    name: "exonum-cryptocurrency:0.12.0"
  
instances:
  xnm-token:
    artifact: cryptocurrency
```

Here, we want to deploy artifact `cryptocurrency`. `cryptocurrency` here is a local artifact definition to be used within config (actual spec is stored inside, see `"exonum-cryptocurrency:0.12.0"`).

If we want to provide a plugin for a runtime, it will be:

```yaml
plugins:
  runtime:
     sample_runtime: our.package.SampleRuntimeSpecLoader
```

if we want also the spec for `cryptocurrency` artifact be loaded with plugin, then plugin section will become the following:

```yaml
plugins:
  runtime:
     sample_runtime: our.package.SampleRuntimeSpecLoader
  artifact:
    cryptocurrency: our.package.CryptocurrencySpecLoader
```

Why is it done that way? At least for `rust` runtime artifact names contain a version. So in order to just bump a version, you will have to update it everywhere (so it's up to N + 1 times -- N for every instance of that artifact in config, and 1 for plugin), with aliases it will be only one place.

And IMO it's just more readable.

>     service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"

Reviewer: Thanks for clarifications, it makes sense!
